Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm

```diff
@@ -0,0 +1,825 @@
+        if (urlMatchesPatternIgnoringWildcardHostPatterns(requestedMatchPattern))
+            return cacheResultAndReturn(PermissionState::RequestedExplicitly);
+
+        if (urlMatchesWildcardHostPatterns(requestedMatchPattern))
+            return cacheResultAndReturn(PermissionState::RequestedImplicitly);
+    }
+
+    if (options.contains(PermissionStateOptions::RequestedWithTabsPermission) && hasPermission(_WKWebExtensionPermissionTabs, tab, options))
+        return PermissionState::RequestedImplicitly;
+
+    return cacheResultAndReturn(PermissionState::Unknown);
+}
+
+WebExtensionContext::PermissionState WebExtensionContext::permissionState(WebExtensionMatchPattern& pattern, _WKWebExtensionTab *tab, OptionSet<PermissionStateOptions> options)
+{
+    if (!pattern.isValid())
+        return PermissionState::Unknown;
+
+    if (pattern.matchesURL(baseURL()))
+        return PermissionState::GrantedImplicitly;
+
+    if (!pattern.matchesAllURLs() && !WebExtensionMatchPattern::validSchemes().contains(pattern.scheme()))
+        return PermissionState::Unknown;
+
+    if (tab && [[m_temporaryTabPermissionMatchPatterns objectForKey:tab] matchesPattern:pattern.wrapper()])
+        return PermissionState::GrantedExplicitly;
+
+    // Access the maps here to remove any expired entries, and only do it once for this call.
+    auto& grantedPermissionMatchPatterns = this->grantedPermissionMatchPatterns();
+    auto& deniedPermissionMatchPatterns = this->deniedPermissionMatchPatterns();
+
+    // First, check for patterns that are specific to certain domains, ignoring wildcard host patterns that
+    // match all hosts. The order is denied, then granted. This makes sure denied takes precedence over granted.
+
+    auto urlMatchesPatternIgnoringWildcardHostPatterns = ^(WebExtensionMatchPattern& pattern) {
+        if (pattern.matchesAllHosts())
+            return false;
+        return pattern.matchesPattern(pattern);
+    };
+
+    for (auto& deniedPermissionEntry : deniedPermissionMatchPatterns) {
+        if (urlMatchesPatternIgnoringWildcardHostPatterns(deniedPermissionEntry.key))
+            return PermissionState::DeniedExplicitly;
+    }
+
+    for (auto& grantedPermissionEntry : grantedPermissionMatchPatterns) {
+        if (urlMatchesPatternIgnoringWildcardHostPatterns(grantedPermissionEntry.key))
+            return PermissionState::GrantedExplicitly;
+    }
+
+    // Next, check for patterns that are wildcard host patterns that match all hosts (<all_urls>, *://*/*, etc),
+    // also checked in denied, then granted order. Doing these wildcard patterns separately allows for blanket
+    // patterns to be set as default policies while allowing for specific domains to still be granted or denied.
+
+    auto urlMatchesWildcardHostPatterns = ^(WebExtensionMatchPattern& pattern) {
+        if (!pattern.matchesAllHosts())
+            return false;
+        return pattern.matchesPattern(pattern);
```

Comment:
Ditto my previous question about this check being necessary.

Also - `pattern.matchesPattern(pattern)` is concerning - which value will pattern have here?

Reply:
Good catch! I think I need to rename the arg!

(Similar to the URL one, the scheme or path might not match.)